### PR TITLE
Ported similarity functions over from DeepQA

### PR DESCRIPTION
**Pull request by matt-gardner**

### CODE REVIEW

File: allennlp/modules/similarity_functions/linear.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    @overrides
+    def forward(self, tensor_1: torch.Tensor, tensor_2: torch.Tensor) -> torch.Tensor:
+        combined_tensors = self._combine_tensors(tensor_1, tensor_2)
+
+        # The '@' operator here is torch.matmul, but that's only available in pytorch-0.2.
+        # TODO(mattg): switch to using torch.matmul when a version of pytorch with it is released.
+        # I think it's more clear and less magical.
+        # Also, broadcasting this simple addition is only available in pytorch-0.2.  When that's
+        # ready, change this back to `(dot_product + self._bias)`.
+        dot_product = combined_tensors @ self._weight_vector
```
Comment (matt-gardner) [inline allennlp/modules/similarity_functions/linear.py]: This multiplication (and the corresponding one in `Bilinear` doesn't handle higher-order tensors.  This is causing three of the tests to fail.  I'm still trying to figure out how to do this properly, though if we could upgrade to pytorch 0.2 somehow, I think the issue would just go away...

File: allennlp/modules/similarity_functions/bilinear.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import torch
+from torch.nn.parameter import Parameter
+
+from allennlp.common import Params
+from allennlp.experiments import Registry
+from allennlp.modules import SimilarityFunction
+
+
+@Registry.register_similarity_function("bilinear")
+class Bilinear(SimilarityFunction):
```
Comment (DeNeutoy) [inline allennlp/modules/similarity_functions/bilinear.py]: Possibly we should use `BilinearSimilarity` and `LinearSimilarity` for these because `torch.nn.Linear` and `torch.nn.Bilinear` already exist.

File: allennlp/modules/similarity_functions/linear.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+from allennlp.experiments import Registry
+from allennlp.modules import SimilarityFunction
+
+
+@Registry.register_similarity_function("linear")
+class LinearSimilarity(SimilarityFunction):
+    """
+    This similarity function performs a dot product between a vector of weights and some
+    combination of the two input vectors, followed by an (optional) activation function.  The
+    combination done is configurable.
```
Comment (DeNeutoy) [inline allennlp/modules/similarity_functions/linear.py]: maybe "used" instead of "done"?

File: allennlp/modules/similarity_functions/linear.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        Described above.
+    activation : ``Callable[[torch.Tensor], torch.Tensor]``, optional (default=``lambda x: x``)
+        An activation function applied after the ``w^T * [x;y] + b`` calculation.  Default is no
+        activation.
+    """
+    def __init__(self,
+                 tensor_1_dim: int,
+                 tensor_2_dim: int,
+                 combination: str = 'x,y',
+                 activation: Callable[[torch.Tensor], torch.Tensor] = lambda x: x) -> None:
```
Comment (DeNeutoy) [inline allennlp/modules/similarity_functions/linear.py]: Can we use a different default activation here and above? Maybe `tanh`? I would expect that a class with an activation argument with a default value to apply some kind of activation.

Comment (matt-gardner) [inline allennlp/modules/similarity_functions/linear.py]: This output is very likely going directly into a softmax, and I wouldn't want to pass those through a `tanh` or most other activation functions first, because then you make it pretty much impossible for the softmax to be peaky.  I think the 'linear' default activation function (same as the default activation for Keras' `Dense` layer) is reasonable here.
